### PR TITLE
crimson/osd: fix the lifetime of Notify during timeouts

### PR DESCRIPTION
**Pull request by rzarzynski**
This fixes a segfault during `LibRadosWatchNotify.WatchNotify2Timeout`.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/osd/watch.cc [hunk 1]
```diff
@@ -210,7 +215,15 @@ void Notify::do_timeout()
   if (complete) {
     return;
   }
+  // it might be that `this` is keep alive only because of the reference
```
Comment (tchaikov) [inline src/crimson/osd/watch.cc:218]: s/keep/kept/

**Review by tchaikov — APPROVED** [review]

File: src/crimson/osd/watch.cc (hunk 1 above)
Comment (rzarzynski) [reply src/crimson/osd/watch.cc]: Fixed. Thanks!

**Review by tchaikov — APPROVED** [review]

File: src/crimson/osd/watch.cc [hunk 2]
```diff
@@ -210,7 +215,15 @@ void Notify::do_timeout()
   if (complete) {
     return;
   }
+  // it might be that `this` is kept alive only because of the reference
+  // a watcher stores and which is being removed by `cancel_notify()`.
+  // to avoid use-after-free we dump up the ref counter with `guard_ptr`.
```
Comment (rzarzynski) [inline src/crimson/osd/watch.cc:220]: Oops, another typo: _dump up_. Fixing.

Comment (rzarzynski) [reply src/crimson/osd/watch.cc]: Fixed.

**Pull request merged**